unistd: add {get,set}domainname

lucab opened the pull request:
This implements `getdomainname(2)` and `setdomainname(2)`, which are
Linux-specific extensions. References:
 * http://man7.org/linux/man-pages/man2/getdomainname.2.html
 * http://man7.org/linux/man-pages/man2/setdomainname.2.html

lucab commented on src/unistd.rs:
```diff
@@ -756,6 +756,36 @@ pub fn gethostname<'a>(buffer: &'a mut [u8]) -> Result<&'a CStr> {
+/// and store it in a provided buffer.  The buffer will be populated with bytes up
+/// to the length of the provided slice including a NUL terminating byte.
+#[cfg(any(target_os = "linux", target_os = "android"))]
+pub fn getdomainname<'a>(buffer: &'a mut [u8]) -> Result<&'a CStr> {
```
I'm not super-happy about the ergonomics of this signature, but I just followed `gethostname` path. Can we rework these to allocate internally and return an owning `CString`?

lucab commented on src/unistd.rs:
```diff
@@ -756,6 +756,36 @@ pub fn gethostname<'a>(buffer: &'a mut [u8]) -> Result<&'a CStr> {
+/// and store it in a provided buffer.  The buffer will be populated with bytes up
+/// to the length of the provided slice including a NUL terminating byte.
+#[cfg(target_os = "linux")]
+pub fn getdomainname<'a>(buffer: &'a mut [u8]) -> Result<&'a CStr> {
```
I'm not super-happy about the ergonomics of this signature, but I just followed gethostname path. Can we rework these to allocate internally and return an owning CString?

asomers replied:
I agree.  I too would rather allocate internally.

asomers commented on src/unistd.rs:
```diff
@@ -756,6 +756,36 @@ pub fn gethostname<'a>(buffer: &'a mut [u8]) -> Result<&'a CStr> {
 
+/// Set the NIS domain name (see
+/// [setdomainname(2)](http://man7.org/linux/man-pages/man2/setdomainname.2.html)).
+#[cfg(target_os = "linux")]
```
Both of these functions are available on FreeBSD too.  And while I haven't checked, I would bet that they're available on all of the BSDs based on their age.

Susurrus replied:
They're also declared on emscripten, so I'd add both on those platforms as well.

lucab replied:
I did a quick check and indeed they seem to be available on all BSD (plus ES) too.

lucab replied:
It looks like libc does not expose them, so https://github.com/rust-lang/libc/issues/868 needs to be fixed first.

Susurrus commented on test/test_unistd.rs:
```diff
@@ -404,3 +404,22 @@ fn test_pipe2() {
 }
+
+#[test]
+fn test_gethostname() {
```
Please put the addition into a separate commit before the rest of these changes.

lucab commented:
Rebased, PTAL. In this iteration:
 * split gethostname test to its own commit
 * changed gethostname and getdomainname to allocate their own buffer
 * added emscripten, reported https://github.com/rust-lang/libc/issues/868 for BSDs (I can tackle this later if nobody else does)

asomers commented:
libc's maintainers probably aren't going to add the definitions themselves, even for reported issues.  Like Nix, they mostly rely on pull requests from their users.  Could you please open a PR for libc?

lucab commented:
This got stale due to lack of time on my side, but I'll eventually get back to this (just not high priority at the moment). In the meanwhile, @gnzlbg is pushing forward the BSD libc fix, kudos to them.

lucab commented:
Rebased to take care of BSD and Apple, PTAL.

Susurrus commented on src/unistd.rs:
```diff
@@ -728,34 +728,33 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
-    let ptr = buffer.as_mut_ptr() as *mut c_char;
-    let len = buffer.len() as size_t;
+pub fn gethostname() -> Result<CString> {
+    const MAXHOSTNAMELEN: usize = 256;
```
We should document why we use the value we use here. Is there some reference example you pulled this from?

lucab replied:
This is `_POSIX_HOST_NAME_MAX` plus the trailing null, from http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html.

I'll rename, add a reference, and leave the trailing byte out of this.

Susurrus commented on src/unistd.rs:
```diff
@@ -728,34 +728,33 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
-        unsafe { CStr::from_ptr(buffer.as_ptr() as *const c_char) }
+        unsafe {
+            // Ensure returned string is always null-terminated
+            *ptr.offset((MAXHOSTNAMELEN - 1) as isize) = 0;
```
Couldn't you instead use a `MAXHOSTNAMELEN` of `255` here and be able to skip this step?

lucab replied:
Yes, that's probably better, I'll do.

lucab commented:
@Susurrus I adjusted and documented `POSIX_HOST_NAME_MAX` as suggested, PTAL.

lucab commented:
Gentle ping for a review.

Susurrus commented on src/unistd.rs:
```diff
@@ -728,34 +728,85 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+pub fn gethostname() -> Result<CString> {
+    // Hostname maximum length as defined by POSIX,
+    // http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html
+    const POSIX_HOST_NAME_MAX: usize = 255;
```
According to this link the max host name length is 64 on Linux, and that's confirmed on my system.  `_POSIX_HOST_NAME_MAX` on my system is 255, however. We should figure out the appropriate values here and get them into `libc` to use them here.

Susurrus commented on src/unistd.rs:
````diff
@@ -728,34 +728,85 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+///
+/// # Examples
+///
+/// ```no_run
````
Can this not run in CI?

asomers replied:
Why use a `CString` here but an `OsStr` for `setdomainname`?  We should be consistent in our choice of C vs OS strings.

lucab replied:
@asomers This follows the split currently in master for the hostname, see `gethostname` vs `sethostname` types. Perhaps it can be left out of this PR? Otherwise, in which direction do you want to resolve this?

Susurrus commented on src/unistd.rs:
```diff
@@ -728,34 +728,85 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+
+/// Set the NIS domain name (see
+/// [setdomainname(2)](http://man7.org/linux/man-pages/man2/setdomainname.2.html)).
+#[cfg(not(target_os = "android"))]
```
We prefer to use whitelists instead of blacklists for listing targets. Please do so here and for `getdomainname`.

Susurrus commented on src/unistd.rs:
```diff
@@ -728,34 +728,85 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+    Errno::result(res).map(|_| {
+        unsafe {
+            // Ensure returned string is always null-terminated
+            *ptr.offset(POSIX_HOST_NAME_MAX as isize) = 0;
```
You shouldn't need to do this because `gethostname` won't overrun the buffer. Same for the other functions.

Susurrus commented on test/test_unistd.rs:
```diff
@@ -400,3 +400,20 @@ fn test_pipe2() {
+#[cfg(not(target_os = "android"))]
+fn test_getdomainname() {
+    let dn = getdomainname();
+    assert!(dn.is_ok(), "Error: {:?}", dn.unwrap_err());
```
Replace this with `dn.expect("")` on the above line.

Susurrus commented on test/test_unistd.rs:
```diff
@@ -400,3 +400,20 @@ fn test_pipe2() {
+#[test]
+fn test_gethostname() {
+    let hn = gethostname();
+    assert!(hn.is_ok(), "Error: {:?}", hn.unwrap_err());
```
Replace this with `hn.expect("")` on the above line.

Susurrus commented on CHANGELOG.md:
```diff
@@ -6,10 +6,14 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 
 ### Added
+- Added `nix::unistd::{getdomainname, setdomainname}`.
```
Please specify which platforms this applies to.

asomers commented on src/unistd.rs:
```diff
@@ -728,34 +728,85 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+#[cfg(not(target_os = "android"))]
+pub fn setdomainname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+    let ptr = name.as_ref().as_bytes().as_ptr() as *const c_char;
+    let len = name.as_ref().len() as namelen_t;
```
I think it would probably be more pedantically correct to do `len = name.as_ref().as_bytes().len()`.

asomers commented on src/unistd.rs:
````diff
@@ -728,34 +728,85 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+/// ```
+#[cfg(not(target_os = "android"))]
+pub fn getdomainname() -> Result<CString> {
+    const MAXDOMAINNAMELEN: usize = 256;
````
According to the man page, `MAXHOSTNAMELEN` applies to the domain name, too.

Susurrus commented:
@lucab Are you planning on coming back to this?

lucab commented:
@Susurrus yes but I need to figure out the story with `HOST_NAME_MAX` Linux and POSIX, and forward to libc.

Susurrus commented:
@lucab Did the `HOST_NAME_MAX` get sorted yet?

asomers commented:
Yes, sysconf sounds like the safest approach.

lucab commented:
Travis is red due to android missing POSIX constants in libc crate, see https://github.com/rust-lang/libc/pull/990.

lucab commented:
Rebased, this is now green. I think I've addressed all comments in reviews so far, but at this point this PR is quite different from its initial version. PTAL.

asomers commented on src/unistd.rs:
```diff
@@ -739,34 +739,97 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+    let buf = vec![0; buf_len as usize];
+    let ptr = unsafe { CString::from_vec_unchecked(buf).into_raw() };
+
+    let res = unsafe { libc::gethostname(ptr, buf_len as usize) };
```
You don't need to convert the buffer to a `CString` first.  You can just do `libc::gethostname(buf.as_mut() as *mut c_char, buf_len as usize`

lucab commented:
@asomers I was a bit unsure about the safety of your comments so I added some idempotency checks to the tests. You are indeed right, so I pushed the rework and the additional testing logic. Everything is green, PTAL.

asomers commented:
Ok, everything looks good to me!  It only needs a squash.  @Susurrus do you have any comments?

lucab commented:
@asomers ack. I squashed one commit out. I'm keeping two commits for the two separate changelog entries. Let me know if you prefer everything together.

Susurrus commented on src/unistd.rs:
````diff
@@ -769,6 +769,60 @@ pub fn gethostname() -> Result<OsString> {
+/// # Examples
+///
+/// ```
+/// use nix::unistd;
````
Please prefix this line with a `#` so they don't render in the final help. You can then also delete the blank line below.

Susurrus commented on src/unistd.rs:
````diff
@@ -769,6 +769,60 @@ pub fn gethostname() -> Result<OsString> {
+/// ```
+/// use nix::unistd;
+///
+/// let domainname_os = unistd::getdomainname().expect("Failed getting domain name");
````
You can reuse the name `domainname` here because of Rust's variable shadowing. Doing this is idiomatic. Please rename this variable to `domainname`.

Susurrus commented on src/unistd.rs:
```diff
@@ -769,6 +769,60 @@ pub fn gethostname() -> Result<OsString> {
+pub fn getdomainname() -> Result<OsString> {
+    // Minimum hostname maximum length as defined by POSIX,
+    // http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html
+    const _POSIX_HOST_NAME_MAX: c_long = 255;
```
Why isn't this constant pulled from `libc`? I didn't see anywhere in the discussion a comment touching on this.

Susurrus commented on src/unistd.rs:
```diff
@@ -728,34 +728,85 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+            *ptr.offset(POSIX_HOST_NAME_MAX as isize) = 0;
+            CString::from_raw(ptr)
+        }
+    })
```
@asomers Does this comment still stand or is it not relevant anymore?

asomers replied:
Unfortunately GitHub has forgotten the context of that comment.  But I think my objection was because he tried to handle that error condition and didn't do it correctly.  Now, he isn't handling any error conditions, which is better.
